### ISSUE TITLE
UploadFile should probably get a reader as an argument

### ISSUE THREAD

**Issue by mitar** [xanzy/go-gitlab#1317]
I find `UploadFile` function signature strange:

```
func (s *ProjectsService) UploadFile(pid interface{}, file string, options ...RequestOptionFunc) (*ProjectFile, *Response, error)
```

I think it would be better to have:

```
func (s *ProjectsService) UploadFile(pid interface{}, r io.Reader, filename string, options ...RequestOptionFunc) (*ProjectFile, *Response, error)
```

In this way you can upload generated file without having to store it to a file first.

This would be a breaking change but I think it would be worth it.

**Comment by svanharmelen** [comment]
As just mentioned in the other thread, I try to follow the GitLab API and examples as close as possible. If you look at https://docs.gitlab.com/ce/api/projects.html#upload-a-file you can see the example there matches this use case exactly.

Yet I can see that for some use cases it would be nice (and potentially easier) to be able to pass a reader instead, so I'm open to change this one to make it more usable.

**Comment by mitar** [comment]
Awesome. I made https://github.com/xanzy/go-gitlab/pull/1318 in this spirit, if you review that, I can extract shared code out and use it for both API calls, I think.

I agree that following examples is useful, but we also want idiomatic Go interfaces in Go package. :-)

**Comment by svanharmelen** [comment]
Fixed by #1318
